### PR TITLE
feat: new cli

### LINKED ISSUE #38: Migrate to clap-derive for cli
**Is your feature request related to a problem? Please describe.**
Currently, the Volt Cli is built using a builder pattern which is fine but the downside is the subcommand matching is not optimal as it is relying on a string for matching 
https://github.com/voltpkg/volt/blob/bd7f714f1a648f07036a28f2c69428acd1e851b7/src/main.rs#L37-L74 

**Describe the solution you'd like**
Using `derive` feature of Clap to build the CLI

---

I would like to contribute to this change. So, I am putting this here to gather feedback before I start working on it :)

### PR DESCRIPTION
**Pull request by numToStr**
This is the initial design of the CLI using clap-derive. Some things might not fit with the new CLI and will require some refactoring. I'll keep the refactoring to a minimum for now (a story for a separate PR), so that we can focus on initial design :)

Closes #38

### CODE REVIEW

File: src/cli.rs
Comment (numToStr) [inline src/cli.rs]: This is the new trait for subcommands.

**Comment by numToStr** [comment]
I am now stuck on the `App` struct https://github.com/voltpkg/volt/blob/6668623fd86c595de856c8ccb8e46fad3cd0b7c3/src/core/utils/app.rs#L29 which is passed into every subcommand. It is not clear to be by just looking at which field is a cli argument or which is not.

**Comment by suptejas** [comment]
`App` stores metadata that each command requires - including command-line args passed into the Volt CLI

**Comment by numToStr** [comment]
Understood. I suppose `package-names` https://github.com/voltpkg/volt/blob/6668623fd86c595de856c8ccb8e46fad3cd0b7c3/src/core/utils/app.rs#L76-L87 is also a cli arg?

---

For now, I just migrate the App struct according to new cli and later we can do the refactoring. Thoughts?

**Comment by suptejas** [comment]
I think we should use the clap `Args` struct to handle arguments - this would work universally (for all commands). Yes, `package-names` is an argument, but it's specific to the `add` command.

**Comment by numToStr** [comment]
> I think we should use the clap Args struct to handle arguments - this would work universally (for all commands).

It'll be easy to just migrate over to derive. 

> Yes, package-names is an argument, but it's specific to the add command.

Then we should add it to `Add` command.

---

The other point of using derive is better help generation and adding proper arguments to every command also reduces confusion about its usage.

**Comment by suptejas** [comment]
I see, let's do it then

**Comment by numToStr** [comment]
`Add` command is migrated to `VoltConfig` (replacing `App`)

You'll be able to compile the app now and see the new help by running `cargo run -- -h`. Also what args are exposed in `App`? I am confused because in `App::initialize` args is just stored , not used.

**Comment by suptejas** [comment]
The app struct just stores information about the current volt configuration (which you have replaced with `VoltConfig`). I'll merge this soon (do rebase when possible to sort out the merge conflicts)

**Comment by numToStr** [comment]
(This is not ready yet, so don't merge. I'll rebase it myself ;))

> The app struct just stores information about the current volt configuration

I meant to say what is exposed to the user? Currently in `VoltConfig`, `--home-dir` is also exposed as a cli arg which doesn't make that much sense. If you could tell me what is exposed to the user now, that'll be great.

**Comment by suptejas** [comment]
Yep, `--home-dir` shouldn't be exposed 😂, a lotta these stuff are actually just environment variables and folder paths that are read later on in the program, perhaps we could use the `State {}` struct to pass on this data around the application.

**Comment by numToStr** [comment]
> perhaps we could use the `State {}` struct to pass on this data around the application.

Yeah, that will make sense. For now, I'll prioritize the derive migration and migrate all the subcommands to `trait VoltCommand` and `VoltConfig`. Later on, we should tackle on the refactoring part.

---

Off topic: Is there any place we can chat about the changes?

**Comment by suptejas** [comment]
Sure, feel free to join the Volt discord server - https://discord.gg/GW9dax9W, we can chat there!

**Comment by numToStr** [comment]
@XtremeDevX This is now ready. Next, I'll try to fully replace the `App` struct with `VoltConfig`.

File: src/commands/node.rs
Comment (ilyvion) [inline src/commands/node.rs]:
```suggestion
            use_windows(self.version).await;
```
This needs to be `self.version`; as is it gives this output on Windows:
```
❯ cargo build
error[E0425]: cannot find value `version` in this scope
   --> src\commands\node.rs:167:25
    |
167 |             use_windows(version).await;
    |                         ^^^^^^^
    |
help: you might have meant to use the available field
    |
167 |             use_windows(self.version).await;
    |                         ~~~~~~~~~~~~
help: consider importing this function
    |
19  | use isahc::version;
    |

For more information about this error, try `rustc --explain E0425`.
error: could not compile `volt` due to previous error
```

**Review by ilyvion — CHANGES_REQUESTED** [review]

File: src/commands/node.rs
Comment (numToStr) [reply src/commands/node.rs]: Ahh, I am on linux so i missed it. Thanks for catching it.

**Comment by numToStr** [comment]
This needs a rebase now

**Comment by ilyvion** [comment]
Yeah, sorry, you picked an unfortunate time to start a rewrite; in the middle of my code cleanup effort. 😅 I'm done for now though, so this should be the last rebase you need to do assuming nobody else comes in with big changes too.

**Comment by numToStr** [comment]
It is also a kind of cleanup 😅. No worries, I'll try to rebase it tomorrow.

**Comment by numToStr** [comment]
@XtremeDevX @alexschrod This is ready. I tried my best to rebase it 😅

**Comment by suptejas** [comment]
Coolio! This is amazing, I'll review changes and merge

**Pull request merged**
